### DISCUSSION TITLE
Scale parameter for linearExtrude

### DISCUSSION THREAD

**Discussion by abejfehr** [jscad/OpenJSCAD.org#1140] [Q&A] [answered]
Similar to my [previous question](https://github.com/jscad/OpenJSCAD.org/discussions/1139), I'm translating a project from OpenSCAD to JSCAD.

[This line](https://github.com/wbu/drawer_organizer/blob/master/drawer_organizer.scad#L236) has an OpenSCAD `linear_extrude` and is given a "scale" parameter, but it doesn't look like `extrudeLinear` accepts that as an option ([here](https://openjscad.xyz/docs/module-modeling_extrusions.html#.extrudeLinear)).

Is there any way for me to scale in an extrusion? Thank you 🙏

**Comment by abejfehr** [comment]
I just realized I may be able to use [extrudeFromSlices](https://openjscad.xyz/docs/module-modeling_extrusions.html#.extrudeFromSlices) for a similar effect, I'll just have to specify the 2D shapes for both ends.

I'd still love to hear alternate answers though!

**Edit:** after reading through `extrudeFromSlices` a few times, it may not behave quite like I expected (I thought it would be like a "loft" function)

**Comment by hrgdavor** [comment]
to be able to suggest sth in jscad i would need an example how that scale parameter affects the result
```
linear_extrude(height=height_linear-gap_top, scale=radius_top_gap/radius_bottom)
```
is the whole object scaled or just one end of the extrusion ?

**Reply by abejfehr** [reply]
Only one end of the extrusion is scaled.

Here's what it looks like currently in JSCAD:
<img width="445" alt="image" src="https://user-images.githubusercontent.com/3917428/187118015-55b4ae1e-ef1b-4bc8-8018-d0557425e89c.png">


And here's what I have in OpenSCAD: 
<img width="504" alt="image" src="https://user-images.githubusercontent.com/3917428/187118081-e9af4f93-640b-4a90-8f95-093ef19b2b41.png">

I'm not sure if it's obvious due to perspective, but the top face is scaled down in the OpenSCAD one

**Comment by hrgdavor** [comment, marked as answer]
there is no such option in extrudeLinear, but a utilit func can be made using `extrudeFromSlices `

```js
const jscad = require('@jscad/modeling')
const {transform, translate, scale} = jscad.transforms
const {circle} = jscad.primitives
const {geom2} = jscad.geometries
const {mat4} = jscad.maths
const {extrudeFromSlices, extrudeLinear, slice} = jscad.extrusions

// extrudeFromSlices is complicated, especially because of usage of slices
function extrudeLinerScaled({height=1,scaleTop=1}, base){
  const callback = (progress,index)=>{
    let localshape = base
    if(index> 0) localshape = scale([scaleTop,scaleTop,scaleTop],base)
    const shape = slice.fromSides(geom2.toSides(localshape))
    const out = slice.transform(mat4.fromTranslation([],[0, 0, index*height]), shape)
    return out
  }
  return extrudeFromSlices({numberOfSlices:2, callback},base)
}

const main = ()=>{
  return [
    translate([20,0,0],extrudeLinear({height:20}, circle({radius:6}))),
    extrudeLinerScaled({height:20, scaleTop:0.5}, circle({radius:6}))
   ]
}

module.exports = {main}
```

![image](https://user-images.githubusercontent.com/2480762/187147290-54b56620-75fb-4dba-9001-e68ffab4455d.png)

**Reply by abejfehr** [reply]
This is pretty much what I was going for, thank you for this!

I couldn't get my 2D geometries to be slices so this is a very helpful reference

**Comment by hrgdavor** [comment]
@abejfehr also if you use discord you can find me there https://discord.com/channels/775984095250612234/914627480512503829

still, it is good to have questions like this here on github for others to find
